File: nautilus_trader/adapters/hyperliquid/parsing.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------
"""
Helper functions for converting Nautilus orders to Hyperliquid format.
"""

import json

from nautilus_trader.model.enums import OrderSide
from nautilus_trader.model.enums import OrderType
from nautilus_trader.model.enums import TimeInForce
from nautilus_trader.model.orders import Order


def extract_asset_from_symbol(symbol: str) -> str:
    """
    Extract asset ID from instrument symbol.
    """
    # Remove venue suffix if present
    if "." in symbol:
        symbol = symbol.split(".")[0]
    # Remove -PERP or -USD suffix
    if symbol.endswith("-PERP"):
        return symbol[:-5]
    if symbol.endswith("-USD"):
        return symbol[:-4]
    return symbol


def order_to_hyperliquid_json(order: Order) -> str:
    """
    Convert a Nautilus order to Hyperliquid order format JSON.

    Parameters
    ----------
    order : Order
        The Nautilus order to convert.

    Returns
    -------
    str
        JSON string representing the Hyperliquid order request.

    """
    # Extract asset from instrument symbol
    asset = extract_asset_from_symbol(str(order.instrument_id.symbol))

    # Determine order side
    is_buy = order.side == OrderSide.BUY

    # Get price (use 0 for market orders as placeholder)
    if order.price is not None:
        price = str(order.price)
    else:
        price = "0"

    # Get size
    size = str(order.quantity)

    # Determine order kind and time-in-force
    if order.order_type == OrderType.MARKET:
        # Market orders are implemented as limit IOC in Hyperliquid
        order_kind = {
            "limit": {
                "tif": "Ioc",
            },
        }
    elif order.order_type == OrderType.LIMIT:
        # Map Nautilus TIF to Hyperliquid TIF
        if order.time_in_force == TimeInForce.GTC:
            tif = "Gtc"
        elif order.time_in_force == TimeInForce.IOC:
            tif = "Ioc"
        elif order.time_in_force == TimeInForce.FOK:
            tif = "Alo"  # Alo = All-or-none (Hyperliquid TIF)
        else:
            tif = "Gtc"  # Default to GTC

        order_kind = {
            "limit": {
                "tif": tif,
            },
        }
    else:
        # For other order types, default to limit GTC
        order_kind = {
            "limit": {
                "tif": "Gtc",
            },
        }

    # Build the order request
    order_request = {
        "asset": asset,
        "isBuy": is_buy,
        "limitPx": price,
        "sz": size,
        "reduceOnly": order.is_reduce_only,
        "orderType": order_kind,
        "cloid": str(order.client_order_id),
    }

    # Return as JSON array (Hyperliquid expects an array of orders)
    return json.dumps([order_request])


def orders_to_hyperliquid_json(orders: list[Order]) -> str:
    """
    Convert multiple Nautilus orders to Hyperliquid orders array JSON.

    Parameters
    ----------
    orders : list[Order]
        List of Nautilus orders to convert.

    Returns
    -------
    str
        JSON string representing the Hyperliquid orders array.

    """
    order_requests = []

    for order in orders:
        # Extract asset from instrument symbol
        asset = extract_asset_from_symbol(str(order.instrument_id.symbol))

        # Determine order side
        is_buy = order.side == OrderSide.BUY

        # Get price (use 0 for market orders as placeholder)
        if order.price is not None:
            price = str(order.price)
        else:
            price = "0"

        # Get size
        size = str(order.quantity)

        # Determine order kind and time-in-force
        if order.order_type == OrderType.MARKET:
            order_kind = {
                "limit": {
                    "tif": "Ioc",
                },
            }
        elif order.order_type == OrderType.LIMIT:
            if order.time_in_force == TimeInForce.GTC:
                tif = "Gtc"
            elif order.time_in_force == TimeInForce.IOC:
                tif = "Ioc"
            elif order.time_in_force == TimeInForce.FOK:
                tif = "Alo"  # Alo = All-or-none (Hyperliquid TIF)
            else:
                tif = "Gtc"

            order_kind = {
                "limit": {
                    "tif": tif,
                },
            }
        else:
            order_kind = {
                "limit": {
                    "tif": "Gtc",
                },
            }

        # Build the order request
        order_request = {
            "asset": asset,
            "isBuy": is_buy,
            "limitPx": price,
            "sz": size,
            "reduceOnly": order.is_reduce_only,
            "orderType": order_kind,
            "cloid": str(order.client_order_id),
        }

        order_requests.append(order_request)

    return json.dumps(order_requests)

File: crates/core/src/python/parsing.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! JSON / string parsing helpers for Python inputs.

use pyo3::{
    prelude::*,
    types::{PyDict, PyList},
};

use super::{to_pykey_err, to_pyvalue_err};

/// Helper function to get a required string value from a Python dictionary.
///
/// # Returns
///
/// Returns the extracted string value or a `PyErr` if the key is missing or extraction fails.
///
/// # Errors
///
/// Returns `PyErr` if the key is missing or value extraction fails.
pub fn get_required_string(dict: &Bound<'_, PyDict>, key: &str) -> PyResult<String> {
    dict.get_item(key)?
        .ok_or_else(|| to_pykey_err(format!("Missing required key: {key}")))?
        .extract()
}

/// Helper function to get a required value from a Python dictionary and extract it.
///
/// # Returns
///
/// Returns the extracted value or a `PyErr` if the key is missing or extraction fails.
///
/// # Errors
///
/// Returns `PyErr` if the key is missing or value extraction fails.
pub fn get_required<T>(dict: &Bound<'_, PyDict>, key: &str) -> PyResult<T>
where
    T: for<'a, 'py> FromPyObject<'a, 'py>,
    for<'a, 'py> PyErr: From<<T as FromPyObject<'a, 'py>>::Error>,
{
    dict.get_item(key)?
        .ok_or_else(|| to_pykey_err(format!("Missing required key: {key}")))?
        .extract()
        .map_err(PyErr::from)
}

/// Helper function to get an optional value from a Python dictionary.
///
/// # Returns
///
/// Returns Some(value) if the key exists and extraction succeeds, None if the key is missing
/// or if the value is Python None, or a `PyErr` if extraction fails.
///
/// # Errors
///
/// Returns `PyErr` if value extraction fails (but not if the key is missing or value is None).
#[inline]
pub fn get_optional<T>(dict: &Bound<'_, PyDict>, key: &str) -> PyResult<Option<T>>
where
    T: for<'a, 'py> FromPyObject<'a, 'py>,
    for<'a, 'py> PyErr: From<<T as FromPyObject<'a, 'py>>::Error>,
{
    match dict.get_item(key)? {
        Some(value) => {
            if value.is_none() {
                Ok(None)
            } else {
                value.extract().map(Some).map_err(PyErr::from)
            }
        }
        None => Ok(None),
    }
}

/// Helper function to get a required value, parse it with a closure, and handle parse errors.
///
/// # Returns
///
/// Returns the parsed value or a `PyErr` if the key is missing, extraction fails, or parsing fails.
///
/// # Errors
///
/// Returns `PyErr` if the key is missing, value extraction fails, or parsing fails.
pub fn get_required_parsed<T, F>(dict: &Bound<'_, PyDict>, key: &str, parser: F) -> PyResult<T>
where
    F: FnOnce(String) -> Result<T, String>,
{
    let value_str = get_required_string(dict, key)?;
    parser(value_str).map_err(|e| to_pyvalue_err(format!("Failed to parse '{key}': {e}")))
}

/// Helper function to get an optional value, parse it with a closure, and handle parse errors.
///
/// # Returns
///
/// Returns `Some(parsed_value)` if the key exists and parsing succeeds, None if the key is missing
/// or if the value is Python None, or a `PyErr` if extraction or parsing fails.
///
/// # Errors
///
/// Returns `PyErr` if value extraction or parsing fails (but not if the key is missing or value is None).
pub fn get_optional_parsed<T, F>(
    dict: &Bound<'_, PyDict>,
    key: &str,
    parser: F,
) -> PyResult<Option<T>>
where
    F: FnOnce(String) -> Result<T, String>,
{
    get_optional::<String>(dict, key)?
        .map(parser)
        .transpose()
        .map_err(|e| to_pyvalue_err(format!("Failed to parse '{key}': {e}")))
}

/// Helper function to get a required `PyList` from a Python dictionary.
///
/// # Returns
///
/// Returns the extracted `PyList` or a `PyErr` if the key is missing or extraction fails.
///
/// # Errors
///
/// Returns `PyErr` if the key is missing or value extraction fails.
pub fn get_required_list<'py>(
    dict: &Bound<'py, PyDict>,
    key: &str,
) -> PyResult<Bound<'py, PyList>> {
    dict.get_item(key)?
        .ok_or_else(|| to_pykey_err(format!("Missing required key: {key}")))?
        .cast_into()
        .map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use std::sync::Once;

    use pyo3::exceptions::{PyKeyError, PyValueError};
    use rstest::rstest;

    use super::*;

    fn ensure_python_initialized() {
        static INIT: Once = Once::new();
        INIT.call_once(Python::initialize);
    }

    #[rstest]
    fn test_get_required_string() {
        ensure_python_initialized();

        Python::attach(|py| {
            let dict = PyDict::new(py);
            dict.set_item("name", "nautilus").unwrap();

            let value = get_required_string(&dict, "name").unwrap();
            let error = get_required_string(&dict, "missing").unwrap_err();

            assert_eq!(value, "nautilus");
            assert!(error.is_instance_of::<PyKeyError>(py));
            assert_eq!(
                error.value(py).to_string(),
                "'Missing required key: missing'"
            );
        });
    }

    #[rstest]
    fn test_get_optional_parsed() {
        ensure_python_initialized();

        Python::attach(|py| {
            let dict = PyDict::new(py);
            dict.set_item("value", "42").unwrap();
            let parsed = get_optional_parsed(&dict, "value", |value| {
                value.parse::<u64>().map_err(|e| e.to_string())
            })
            .unwrap();
            let missing = get_optional_parsed(&dict, "missing", |value| {
                value.parse::<u64>().map_err(|e| e.to_string())
            })
            .unwrap();

            dict.set_item("value", py.None()).unwrap();
            let none = get_optional_parsed(&dict, "value", |value| {
                value.parse::<u64>().map_err(|e| e.to_string())
            })
            .unwrap();

            dict.set_item("value", "invalid").unwrap();
            let error =
                get_optional_parsed::<u64, _>(&dict, "value", |_| Err("not a number".to_string()))
                    .unwrap_err();

            assert_eq!(parsed, Some(42));
            assert_eq!(missing, None);
            assert_eq!(none, None);
            assert!(error.is_instance_of::<PyValueError>(py));
            assert_eq!(
                error.value(py).to_string(),
                "Failed to parse 'value': not a number"
            );
        });
    }
}
